src/components/DepositWidget/index.tsx

```diff
@@ -72,6 +99,152 @@ const DepositWidget: React.FC = () => {
+    }, HIGHLIGHT_TIME)
+  }
+
+  async function _deposit(amount: BN, tokenAddress: string): Promise<void> {
+    try {
+      const { symbol, decimals } = _getToken(tokenAddress)
+      log(`Processing deposit of ${amount} ${symbol} from ${userAddress}`)
+      const result = await depositApi.deposit(userAddress, tokenAddress, amount, txOptionalParams)
+      log(`The transaction has been mined: ${result.receipt.transactionHash}`)
+
+      if (mounted.current) {
+        _updateToken(tokenAddress, ({ depositingBalance, walletBalance, ...otherParams }) => {
+          return {
+            ...otherParams,
+            depositingBalance: depositingBalance.add(amount),
+            walletBalance: walletBalance.sub(amount),
+            highlighted: true,
+          }
+        })
+        _clearHighlight(tokenAddress)
+      }
+
+      toast.success(`Successfully deposited ${formatAmount(amount, decimals)} ${symbol}`)
+    } catch (error) {
+      console.error('Error depositing', error)
+      toast.error(`Error depositing: ${error.message}`)
+    }
+  }
+
+  async function _requestWithdraw(amount: BN, tokenAddress: string): Promise<void> {
+    const { symbol, decimals } = _getToken(tokenAddress)
+    try {
+      log(`Processing withdraw request of ${amount} ${symbol} from ${userAddress}`)
+
+      const result = await depositApi.requestWithdraw(userAddress, tokenAddress, amount, txOptionalParams)
+      log(`The transaction has been mined: ${result.receipt.transactionHash}`)
+
+      if (mounted.current) {
+        _updateToken(tokenAddress, otherParams => {
+          return {
+            ...otherParams,
+            withdrawingBalance: amount,
+            claimable: false,
+            highlighted: true,
+          }
+        })
+        _clearHighlight(tokenAddress)
+      }
+
+      toast.success(`Successfully requested withdraw of ${formatAmount(amount, decimals)} ${symbol}`)
+    } catch (error) {
+      console.error('Error requesting withdraw', error)
+      toast.error(`Error requesting withdraw: ${error.message}`)
+    }
+  }
+
+  async function _claim(tokenAddress: string): Promise<void> {
+    const { withdrawingBalance, symbol, decimals } = _getToken(tokenAddress)
+    try {
+      console.debug(`Starting the withdraw for ${formatAmountFull(withdrawingBalance, decimals)} of ${symbol}`)
+      _updateToken(tokenAddress, otherParams => {
+        return {
+          ...otherParams,
+          claiming: true,
+        }
+      })
+      const result = await depositApi.withdraw(userAddress, tokenAddress, txOptionalParams)
+
+      if (mounted.current) {
+        _updateToken(tokenAddress, ({ exchangeBalance, walletBalance, ...otherParams }) => {
+          return {
+            ...otherParams,
+            claiming: false,
+            exchangeBalance: exchangeBalance.sub(withdrawingBalance),
+            withdrawingBalance: ZERO,
+            claimable: false,
+            walletBalance: walletBalance.add(withdrawingBalance),
```

Comment:
calls to `_updateToken` that change balances make this otherwise api-controlled component uncontrolled. The way I see it -- the source of truth for balances comes from `useTokenBalances`, which will get updated when `depositApi.withdraw` gets its tx processed, or just when blockchain state is changed. But when you setState with new balances, best case -- it acts as an optimistic update, worst - due to possible race conditions (in async function), balances get subtracted twice.

Imho, get balances from api only, make requests to the api, get new balances from the api -- simpler and better.

Reply:
Very good point. I agree, however `useTokenBalances` doesn't listen for changes for now. I'll create an issue for this, as for now, there wouldn't be racing conditions since the balances are only updated using _updateToken. 

After we do https://github.com/gnosis/dex-react/issues/96 there might be a race condition, super unlikely.

Anyways, at some point we should react better to the state change and remove this imperative state updates in favour of responsive data flowing down.

WDYT?

Reply:
Also, this is not part of this PR since this is just moving the logic up. So whatever we do, better to address it in a separate PR.

Reply:
That's perfectly good. It will be listening eventually.